### ISSUE TITLE
Wiki Octave installation on MacOS with Spack

### ISSUE THREAD

**Issue by LDAmorim** [spack/spack#15986]
Hi all,

I noticed that Octave package installation in macOS is possible with spack, so I thought it would be interesting to add that possibility to the wiki page:

https://wiki.octave.org/Octave_for_macOS

I left there a draft and I would like to ask you if you want to check it out and correct / update what I wrote.

ps: If you have a working version of Octave in macOS Catalina installed via spack, you can share that information, so that other users can know all working versions.

Cheers

**Comment by adamjstewart** [comment]
Looks good to me. @davydden might be using Octave on macOS, but I'm not sure who else is.

**Comment by adamjstewart** [comment]
@mtmiller may also be interested in this conversation, as he's maintaining the Octave package in Spack for us.

**Comment by davydden** [comment]
it's been a while since I  built Octave on macOS with spack, but that used to work.

**Comment by mtmiller** [comment]
I know nothing about macOS, but what you've added to the wiki looks good to me.

**Comment by Noerr** [comment]
Wow.  Timing couldn't be better.   I happened to find your [wiki page](https://wiki.octave.org/Octave_for_macOS) edit this morning through Google search and give it a go.  Eight hours later, I don't have a working solution.
Now I am stumbling upon [this->issue](https://github.com/spack/spack/issues/15986).  @LDAmorim , Did the basic steps `spack install octave` work for you?

Starting from a clean slate macOS El Capitan (10.14.6) with XCode 11.3.

- Installed Spack v0.14 
  - `git clone https://github.com/spack/spack.git`
  - `git checkout releases/v0.14`
- `spack install gcc@8.3.0`
  - The default gcc 9.2.0 introduced worse issues with undefined symbols inside the gcc C++ header install, unrelated to Octave. I decided to back down a whole version and start over.
- `spack install octave %gcc@8.3.0 `
- **Fundamental issue:** An octave source file includes an X11 header, but X11_INCFLAGS are not supplied to the compile command.  X11_INCFLAGS is set to `X11_INCFLAGS = /usr/X11/include` in the Makefile (generated through Spack and autotools behavior).  this is a working pre-existing X11 install that pre-existed today's efforts. It seems autoconf found it.
    - Root error long into the spack install:
    - `libinterp/corefcn/cdisplay.c:34:10: fatal error: X11/Xlib.h: No such file or directory`
    - The source file line is this:  `#include <X11/Xlib.h>`
    - And that file does indeed exist at: `/usr/X11/include/X11/Xlib.h` but I believe the compile line is missing a `-I/usr/X11/include`



I also notice that the default Octave Spack configuration does *not* enable gnuplot variant.  Ultimately I'm going to want to plot things.   `spack instal octave gnuplot=True`
The Spack DAG explodes for this variant - building tons of system software and libraries.

**Comment by Noerr** [comment]
I'm relatively new to Spack but trying to learn more.   I might be able to troubleshoot matters further if I could coax more debugging output from the build process.
Can anyone suggest how to ultimately get the actual build commands issued to the terminal or to the build log?

Perhaps make V=1. VERBOSE=1 make -d ??? I don't know.  But I can't even experiment because I don't understand how to get something like `spack install` to trigger that kind of more verbose output from make.

**Comment by LDAmorim** [comment]
Hi @Noerr ,
It's nice that you decided to test it! Sorry it didn't work for you (yet).
Yes, when I posted this I had gotten Octave to work without a problem, but I was using Catalina 10.14.3.
I currently have access to a Catalina only, but maybe if you share the build output and error files (you should find an octave folder in another one in `/var/folers/...`).
Actually maybe to make diagnosing your issue easier, could you please try to open a separate issue, just to follow the spack GitHub issue template that lists all the information that is usually useful for debugging it?
Thanks, Diana

**Comment by Noerr** [comment]
Thanks for the encouragement @LDAmorim .  I'm excited the build works for you.
Can I ask: What is your output for `$spack compilers`  ?
octave requires openblas.  openblas requires a Fortran compiler, which is not part of Xcode.  I'd like to recreate your configuration.

I borrowed another family MacBook running Catalina.  It is essentially a blank slate with no other development software.  I can't reproduce the steps in your wiki article because it does not address the compiler installation / selection.

**Comment by Noerr** [comment]
I did open one [issue](https://github.com/spack/spack/issues/16053).  I'll continue to do so if I isolate specific problems.

**Comment by LDAmorim** [comment]
That is correct. I added the gfortran to the clang compiler configuration of spack, by changing the two lines inside clang compiler in. the file `~/.spack/darwin/compilers.yaml`:
```
      f77: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.3-apple/gcc-9.3.0-xmd4ztofmwdhv56ho452xzv6uz7nbwyn/bin/gfortran
      fc: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.3-apple/gcc-9.3.0-xmd4ztofmwdhv56ho452xzv6uz7nbwyn/bin/gfortran
```
You would have to link them to your own gfortran path, you can copy it from the gcc compiler lines.

Thanks! I will check out the issue and try to follow up and test it myself.

**Comment by Noerr** [comment]
Sorry, which clang compiler are you using?  
I tried the fresh build on the Catalina laptop.  

- `spack install gcc`
- pointed my only `~/.spack/darwin/compilers.yaml` compiler definition (Xcode`s clang) to also use `f77` and `fc` from the new gcc path (looked very similar to your two lines above)
- `spack install openblas`

The openblas compile failed due to an internal clang error:
`cannot emit physreg copy instruction`
Poking around for similar mentions led me to another @LDAmorim  finding issue [16028](https://github.com/spack/spack/issues/16028).

I'm not clear if you're avoiding openblas by utilizing MKL or if you installed another llvm/clang that does not hit the clang internal error.

I'm going to try `spack install llvm` overnight and see if that opens up opportunities tomorrow.

**Comment by LDAmorim** [comment]
Hi, sorry my spack environment changed (which means I tested multiple combinations of compilers and packages) since then, but I think I was using MKL for that reason.
I hope it worked for you with llvm.

**Comment by Noerr** [comment]
Sadly the llvm did not work either.
```
$ spack install llvm
$ spack compiler add $(spack location -i llvm)
$ spack edit compilers # insert the paths from gcc for f77 and fc in clang@9.0.1 spec
$ spack install octave %clang@9.0.1
...
```
I get to the octave build again. Then linker undefined symbols `err_disabled_feature`, `octave_dld_function::create`, `check_version`.


I'm out of ideas.  I don't think the [octave wiki spack steps](https://wiki.octave.org/Octave_for_macOS) are representative.

**Comment by LDAmorim** [comment]
Could you try checking out to the develop branch of spack?

**Comment by Noerr** [comment]
I tried several tests with the develop branch today and they all ended up in the same errors.  The only real difference is that I can now run 
```
$ spack debug report
* **Spack:** 0.14.1-649-94aa368af
* **Python:** 3.7.3
* **Platform:** darwin-catalina-ivybridge
```

I started completely over.

- new Terminal session
- `rm -rf ~/.spack`
- `git pull ...spack`
- `source share/spack/setup-env.sh`


```
$ spack install octave  # octave fails - no Fortran compiler

$ spack install gcc # succeeds
$ # modified the only Apple clang compiler config to point fc and f77 at the gcc/gfortran
$ spack install octave # openblas build fails with [Cannot emit physreg copy instruction](https://github.com/spack/spack/issues/16028)

# so try and use compiler GCC
$ spack compiler add $(spack location -i gcc)
$ spack install octave %gcc # octave build linker fails due to undefined symbols similar to my previous comment.

Then linker undefined symbols:
err_disabled_feature
octave_dld_function::create
check_version

```
